phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java

```diff
@@ -581,92 +601,135 @@ private boolean areMutationsInSameTable(Table targetHTable, Region region) {
                 region.getTableDescriptor().getTableName().getName()) == 0);
     }
 
+    @Override
+    public InternalScanner preFlush(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+            InternalScanner scanner, FlushLifeCycleTracker tracker) throws IOException {
+        if (!isPhoenixTableTTLEnabled(c.getEnvironment().getConfiguration())) {
+            return scanner;
+        } else {
+            return User.runAsLoginUser(new PrivilegedExceptionAction<InternalScanner>() {
+                @Override public InternalScanner run() throws Exception {
+                    String tableName = c.getEnvironment().getRegion().getRegionInfo().getTable()
+                            .getNameAsString();
+                    long maxLookbackInMillis =
+                            UngroupedAggregateRegionObserver.getMaxLookbackInMillis(
```

Comment:
This guarantees eventual consistency and the static map is filled up by pre-compact hook. So that means once DDL to change max lookback for a table is run, new value will start getting used on region server (updated in static map) when pre-compact hook on that table is called. So, I have following doubts:
- Is there a set frequency within which pre-compact will get called at least once for each table on each region server hosting regions of that table?
- For short duration within the max lookback map just after changing max lookback, value of max lookback for table and index/view and view index might not match, is that fine?

Reply:
Yes, you are right that this solution here is eventual consistent. The precompact is called for minor and major compaction. Yes, major compaction is done periodically typically once a week. However, the compaction policy may decide to skip it.  Minor compaction is supposed to run whenever the number of HFiles exceeds the configured threshold. If the eventual consistency is fine for data table regions, then eventual consistency between data table region and index from the max lookback perspective should be also fine. However, I think we need a better solution than this such that we need to populate and invalidate this cache whenever the max lookback config changes. The solution here is a stopgap. Hopefully, we will implement a better solution before 5.3.  Please note we can continue to configure "phoenix.max.lookback.age.seconds" at the table level to using the Phoenix alter table command to get much better eventual consistency meanwhile.